include/boost/url/impl/url.ipp

```diff
@@ -2042,26 +2025,157 @@ resolve(
 //
 //------------------------------------------------
 
-url&
+void
 url::
-normalize()
+normalize_octets_impl(
+    int id,
+    grammar::lut_chars const& cs) noexcept
 {
-#if 0
-    // scheme
+    char* it = s_ + offset(id);
+    char* end = s_ + offset(id + 1);
+    char buf = 0;
+    char* dest = it;
+    while (it < end)
     {
-        char* it = s_
-            + offset(id_scheme);
-        auto last = it + offset(id_user);
-        if(it < last)
+        if (*it != '%')
+        {
+            *dest = *it;
+            ++it;
+            ++dest;
+            continue;
+        }
+        if (end - it < 3)
+            break;
+
+        // decode unreserved octets
+        pct_decode_unchecked(
+            &buf,
+            &buf + 1,
+            string_view(it, 3));
+        if (cs(buf))
         {
-            --last;
-            while(it != last)
+            *dest = buf;
+            it += 3;
+            ++dest;
+            continue;
+        }
+
+        // uppercase percent-encoding triplets
+        ++it;
+        *it = grammar::ascii_toupper(*it);
+        ++it;
+        *it = grammar::ascii_toupper(*it);
+        ++it;
+        dest += 3;
+    }
+    if (it != dest)
+    {
+        std::size_t diff = it - dest;
+        std::size_t s = len(id) - diff;
+        shrink_impl(id, s);
```

Comment:
Why aren't we just calling `resize(id,s)`?

Reply:
Hmm `s` as an integer is confusing can we please use `n` for sizes and `s` for strings or string views? I started to look to `shrink_impl` to understand why it was being passed a string...

Reply:
> Why aren't we just calling resize(id,s)?

Resize also calls `shrink_impl`.  `shrink_impl` is just the snippet from `resize_impl` when the new size is smaller.

The difference is only `resize_impl` would (i) redundantly check if we need to increase or decrease the size, (ii) make `noexcept` impossible if we ever want to, and (iii) semantics, to express directly in code that normalization never allocates.